stm32-modules/reference-module/firmware/host_comms_task/freertos_comms_task.cpp

```diff
@@ -0,0 +1,165 @@
+/*
+ * firmware-specific functions, data, and hooks for host comms control
+ */
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+
+#include "FreeRTOS.h"
+#include "firmware/firmware_tasks.hpp"
+#include "firmware/freertos_message_queue.hpp"
+#include "firmware/usb_hardware.h"
+#include "hal/double_buffer.hpp"
+#include "reference-module/host_comms_task.hpp"
+#include "reference-module/messages.hpp"
+#include "task.h"
+
+/** Sadly this must be manually duplicated from usbd_cdc.h */
+constexpr size_t CDC_BUFFER_SIZE = 512U;
+
+// Store any static data for USB comms
+struct CommsTaskFreeRTOS {
+    double_buffer::DoubleBuffer<char, CDC_BUFFER_SIZE * 4> rx_buf;
+    double_buffer::DoubleBuffer<char, CDC_BUFFER_SIZE * 4> tx_buf;
+    char *committed_rx_buf_ptr;
+};
+
+static auto cdc_init_handler() -> uint8_t *;
+static auto cdc_deinit_handler() -> void;
+// NOLINTNEXTLINE(readability-named-parameter)
+static auto cdc_rx_handler(uint8_t *, uint32_t *) -> uint8_t *;
+
+namespace host_comms_control_task {
+
+enum class Notifications : uint8_t {
+    INCOMING_MESSAGE = 1,
+};
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+static tasks::FirmwareTasks::HostCommsQueue
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    _comms_queue(static_cast<uint8_t>(Notifications::INCOMING_MESSAGE),
+                 "Comms Queue");
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+static CommsTaskFreeRTOS _local_task = {
+    .rx_buf = {}, .tx_buf = {}, .committed_rx_buf_ptr = nullptr};
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+static auto _top_task = host_comms_task::HostCommsTask(_comms_queue, nullptr);
+
+auto run(tasks::FirmwareTasks::QueueAggregator *aggregator) -> void {
+    auto *local_task = &_local_task;
+    auto *top_task = &_top_task;
+
+    auto *handle = xTaskGetCurrentTaskHandle();
+
+    _comms_queue.provide_handle(handle);
+    top_task->provide_aggregator(aggregator);
+    aggregator->register_queue(_comms_queue);
+
+    usb_hw_init(&cdc_rx_handler, &cdc_init_handler, &cdc_deinit_handler);
+    usb_hw_start();
+    local_task->committed_rx_buf_ptr = local_task->rx_buf.committed()->data();
+    _comms_queue.set_ready();
+    while (true) {
+        char *tx_end =
+            top_task->run_once(local_task->tx_buf.accessible()->begin(),
+                               local_task->tx_buf.accessible()->end());
+        if (!top_task->may_connect()) {
+            usb_hw_stop();
+        } else if (tx_end != local_task->tx_buf.accessible()->data()) {
+            local_task->tx_buf.swap();
+            usb_hw_send(
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                reinterpret_cast<uint8_t *>(
+                    local_task->tx_buf.committed()->data()),
+                tx_end - local_task->tx_buf.committed()->data());
```

Comment:
from bitter recent experience add a `vTaskDelay(1);` after the call to `usb_hw_send()`, because otherwise if multiple messages are in the host comms queue to send, you can end up swapping the to-send buffer back _out_ of the usb hardware before the next 125us tick hits and the data actually goes to the wire.

Reply:
wow thats insane.